vermouth/tests/test_average_beads.py

```diff
@@ -0,0 +1,133 @@
+        (1, {'mass': 1.2, 'not mass': 2.2, 'position': np.array([2, 3, 4], dtype=float),}),
+        (2, {'mass': 1.3, 'not mass': 2.3, 'position': np.array([3, 4, 5], dtype=float),}),
+    ))
+    mol.add_node(0, **{
+        "graph": subgraph,
+        "mapping_weights": {0: 1, 1: 2, 2: 3},
+        "target mass": np.array([2.378378378, 3.378378378, 4.378378378]),
+        "target not mass": np.array([2.358208955, 3.358208955, 4.358208955]),
+        "target None": np.array([2.333333333, 3.333333333, 4.333333333]),
+        "target False": np.array([2.333333333, 3.333333333, 4.333333333]),
+    })
+
+
+    subgraph = nx.Graph()
+    subgraph.add_nodes_from((
+        (0, {'mass': 1.2, 'not mass': 2.2, 'position': np.array([2, 3, 4], dtype=float),}),
+        (1, {'mass': 1.3, 'not mass': 2.3, 'position': np.array([3, 4, 5], dtype=float),}),
+    ))
+    mol.add_node(1, **{
+        "graph": subgraph,
+        "mapping_weights": {0: 2, 1: 3},
+        "target mass": np.array([2.619047619, 3.619047619, 4.619047619]),
+        "target not mass": np.array([2.610619469, 3.610619469, 4.610619469]),
+        "target None": np.array([2.6, 3.6, 4.6]),
+        "target False": np.array([2.6, 3.6, 4.6]),
+    })
+
+    return mol
+
+@pytest.fixture(params=(None, 'mass', 'not mass'))
+def mol_with_variable(request, mol_with_subgraph):
+    """
+    Build a mock molecule with a mock force field declaring 'center_weight'.
+    """
+    weight = request.param
+
+    class MockForceField(object):
+        pass
+
+    ff = MockForceField()
+    ff.variables = {'center_weight': weight}
+
+    mol_with_subgraph.force_field = ff
+    return mol_with_subgraph
+
+
+@pytest.mark.parametrize('weight', (None, 'mass', 'not mass'))
+def test_do_average_bead(mol_with_subgraph, weight):
+    """
+    Test normal operation of :func:`average_beads.do_average_bead`.
+    """
+    result_mol = average_beads.do_average_bead(
+        mol_with_subgraph, ignore_missing_graphs=False, weight=weight,
+    )
+    target_key = 'target {}'.format(weight)
+    target_positions = np.stack([node[target_key] for node in mol_with_subgraph.nodes.values()])
+    positions = np.stack([node['position'] for node in mol_with_subgraph.nodes.values()])
+    assert np.allclose(positions, target_positions)
+
+
+@pytest.mark.parametrize('weight', ('mass', 'not mass'))
+def test_shoot_weight(mol_with_subgraph, weight):
+    """
+    Test that :func:`average_beads.do_average_bead` fails if a weight is missing.
+    """
+    del mol_with_subgraph.nodes[0]['graph'].nodes[1][weight]
+    with pytest.raises(KeyError):
+        average_beads.do_average_bead(
+            mol_with_subgraph, ignore_missing_graphs=False, weight=weight,
+        )
+
+
+def test_shoot_graph(mol_with_subgraph):
+    """
+    Test that :func:`average_beads.do_average_bead` fails if a subgraph is missing.
+    """
+    del mol_with_subgraph.nodes[1]['graph']
+    with pytest.raises(ValueError):
```

Comment:
Regarding reconstructing beads/atoms, missing sidechains, and generating coordinates; I'm not sure this is a good idea.

Reply:
This behavior did not change: if there is no underlying graph and the `ignore_missing_graphs` keyword argument of `do_average_bead` is `False`, then you get a `ValueError`.